### ISSUE TITLE
Training steps compared against wrong value when determining when to end training when using Gradient Accumulation.

### ISSUE THREAD

**Issue by breakcore2** [d8ahazard/sd_dreambooth_extension#627]
`
Commit hash: d06592267c745b4732026c4e0c499c9a4b3900a1
Installing requirements for Web UI

#######################################################################################################
Initializing Dreambooth
If submitting an issue on github, please provide the below text for debugging purposes:

Python revision: 3.10.7 (tags/v3.10.7:6cc6b13, Sep  5 2022, 14:08:36) [MSC v.1933 64 bit (AMD64)]
Dreambooth revision: c5cb58328c555ac27679422b1da940a9b19de6f2
SD-WebUI revision: d06592267c745b4732026c4e0c499c9a4b3900a1

[+] bitsandbytes version 0.35.0 installed.
[+] diffusers version 0.10.2 installed.
[+] transformers version 4.25.1 installed.
[+] xformers version 0.0.14.dev0 installed.
[+] torch version 1.12.1+cu113 installed.
[+] torchvision version 0.13.1+cu113 installed.
`

**Have you read the [Readme](https://github.com/d8ahazard/sd_dreambooth_extension#readme)?**
Yes.

**Have you completely restarted the stable-diffusion-webUI, not just reloaded the UI?**
Yes

**Have you updated Dreambooth to the latest revision?**
Yes

**Have you updated the Stable-Diffusion-WebUI to the latest version?**
No. This has nothing to do with the Stable-Diffusion-WebUI
**No, really. Please save us both some trouble and update the SD-WebUI and Extension and restart before posting this. Reply 'OK' Below to acknowledge that you did this.**
OK

**Describe the bug**
training_complete is comparing global_step against actual_train_steps to determine if the training is done but actual_train_steps does not represent (number of training images) * (desired epoch) when gradient accumulation != 1 so training is completed earlier than desire based on user input value of epoch size and dataset size
train_dreambooth.py:line 1069
training_complete = global_step >= actual_train_steps or shared.state.interrupted

global_step appears to represent the number of images seen.
train_dreambooth.py:line 1108
global_step += args.train_batch_size # this is done per training example per epoch where a training example is batch_size number of images.

Meanwhile, actual_train_steps represents number of images in training set * desired epoch amount / gradient accumulation steps.
train_dreambooth.py:line 823
actual_train_steps = max_train_steps * args.train_batch_size
train_dreambooth.py:line 820
max_train_steps = args.num_train_epochs * num_update_steps_per_epoch
train_dreambooth.py:line 818 (removed the math.ceil for clarity here)
num_update_steps_per_epoch = len(train_dataloader) / args.gradient_accumulation_steps
len(train_dataloader) = number of images in data set / args.train_batch_size

So some math:
actual_train_steps = max_train_steps * args.train_batch_size # line 823
                             = args.num_train_epochs * num_update_steps_per_epoch * args.train_batch_size # substitute value of max_train_steps
                             = args.num_train_epochs * (len(train_dataloader) / args.gradient_accumulation_steps) * args.train_batch_size # substitute value of num_update_steps_per_epoch
                             = args.num_train_epochs * ((number of images in data set / args.train_batch_size) / args.gradient_accumulation_steps) * args.train_batch_size # substitute value of len(train_dataloader)
                             = args.num_train_epochs * (number of images in data set / (args.train_batch_size * args.gradient_accumulation_steps)) * args.train_batch_size # simplify the division
                            = args.num_train_epochs * (number of images in data set / args.gradient_accumulation_steps) # args.train_batch_size cancel out
                            = (args.num_train_epochs * number of images in data set) / args.gradient_accumulation_steps # rearrange brackets

We see here that actual_train_steps is the total number of images that would be seen divided by the gradient accumulation step count.
This is causing training to end a lot earlier than intended. If I want to run 10 epochs with 10 images, I expect to go through 10 * 10 global_steps but if GA = 2, I would only go through 10 * 10/2.

**Provide logs**
(you can see here that 400 = 100 * 1152/88 and the it/s is the same I would get if I had GA=1 which means it isnt like global_step is only incrementing after seeing all images of a gradient accumulation step)

***** Running training *****
  Num examples = 1152
  Num batches each epoch = 1152
  Num Epochs = 100
  Instantaneous batch size per device = 1
  Total train batch size (w. parallel, distributed & accumulation) = 288
  Gradient Accumulation steps = 288
  Total optimization steps = 400
  Actual steps: 400
   Training settings: CPU: False Adam: True, Prec: fp16, Grad: True, TextTr: False EM: False, LR: 0.0008 LORA:True
 Allocated: 3.9GB
 Reserved: 4.0GB

Steps:   0%|                                                                                                                                                               | 0/400 [00:00<?, ?it/s, loss=0.00281, lr=0.0008, vram=3.9/4.3GB] Step 0 completed.
 Allocated: 3.9GB
 Reserved: 4.3GB

Steps:   1%|█▉                                                                                                                                                       | 5/400 [00:04<07:20,  1.11s/it, loss=0.109, lr=0.0008, vram=3.9/4.3GB] Step 5 completed.
 Allocated: 3.9GB
 Reserved: 4.3GB

Steps:  10%|██████████████▊                                                                                                                                         | 39/400 [00:37<06:36,  1.10s/it, loss=0.166, lr=0.0008, vram=3.9/4.3GB]

**Environment**

What OS?
Windows 10
If Windows - WSL or native?
native
![example1](https://user-images.githubusercontent.com/121475899/209631862-f8d518d0-f66d-41ef-ac8b-ae48065d299b.png)

What GPU are you using?
1080
**Screenshots/Config**

**Comment by d8ahazard** [comment]
Fixed with https://github.com/d8ahazard/sd_dreambooth_extension/commit/15244fef18c35864be6f6c2e1c0237dadc5415c1.
